course03/project01/balint/application.cc

```diff
@@ -0,0 +1,257 @@
+      ;
+      mGameOver = true;
+    }
+  mBallPos = advanceBall();
+}
+vec2 BasicApp::advanceBall() { return mBallPos + mBallVel; }
+
+void BasicApp::setup()
+{
+  // init
+
+  float halfHeight = height / 2;
+  float halfWidth = width / 2;
+  float mLeftBatPos = halfHeight;
+  float mRightBatPos = halfHeight;
+  mBallPos = vec2(halfWidth, halfHeight);
+  mBallVel = vec2(1, 1);
+
+  mLeftBarrier = Rectf(0, 0, mBarrierWidth, height);
+  mRightBarrier = Rectf(width - mBarrierWidth, 0, width, height);
+  mTopBarrier = Rectf(0, 0, width, mBarrierWidth);
+  mBottomBarrier = Rectf(0, height - mBarrierWidth, width, height);
+
+  mLeftBat = Rectf(mBarrierWidth, mLeftBatPos - 0.5 * mBatLength, 2 * mBarrierWidth, mLeftBatPos + 0.5 * mBatLength);
+  mRightBat = Rectf(width - 2 * mBarrierWidth, mRightBatPos - 0.5 * mBatLength, width - mBarrierWidth,
+		    mRightBatPos + 0.5 * mBatLength);
+}
+
+void BasicApp::moveLeftBat(bool up_p)
+{
+  vec2 move = vec2(0, up_p ? -1 : 1);
+  // std::cout<<"MOving left bat up: "<<up_p<<std::endl;
+  if ((mLeftBat.getCenter() + move).x > 0 && (mLeftBat.getCenter() + move).x < height)
+    {
+      mLeftBat.offset(move);
+    }
+}
+void BasicApp::moveRightBat(bool up_p)
+{
+  vec2 move = vec2(0, up_p ? -20 : 20);
+  // std::cout<<"MOving right bat up: "<<up_p<<std::endl;
+  mRightBat.offset(move);
+}
+
+void BasicApp::keyDown(KeyEvent event)
+{
+  if (event.getChar() == 'f')
+    {  // Toggle full screen when the user presses the 'f' key.
+      setFullScreen(!isFullScreen());
+    }
+  else if (event.getCode() == KeyEvent::KEY_ESCAPE)
+    {
+      // Exit full screen, or quit the application, when the user presses the ESC key.
+      if (isFullScreen())
+	setFullScreen(false);
+      else
+	quit();
+    }
+  else if (event.getCode() == KeyEvent::KEY_UP)
+    {
+      moveRightBat(true);
+    }
+  else if (event.getCode() == KeyEvent::KEY_DOWN)
+    {
+      moveRightBat(false);
+    }
+}
+void BasicApp::drawSurroundings()
+{
+  gl::color(Color::white());
+  gl::drawSolidRect(mLeftBarrier);
+  gl::color(Color(1, 0, 0));  // red
+  gl::drawSolidRect(mTopBarrier);
+  gl::color(Color(0, 1, 0));  // green
+  gl::drawSolidRect(mBottomBarrier);
+  gl::color(Color(0, 0, 1));  // blue
+  gl::drawSolidRect(mRightBarrier);
+}
+void BasicApp::drawBall()
+{
+  gl::color(Color::white());
+  gl::drawSolidCircle(mBallPos, mBallRadius);
+}
+
+Rectf getBallBoundingRect(vec2& ballPos, float ballRadius)
+{
+  return Rectf(ballPos.x - ballRadius, ballPos.y - ballRadius, ballPos.x + ballRadius, ballPos.y + ballRadius);
+}
+bool BasicApp::ballCollidesWithRect(Rectf& rect)
+{
+  vec2 NextBallPos = advanceBall();
+  Rectf RectBoundingNextBallPos = getBallBoundingRect(NextBallPos, mBallRadius);
+  // std::cout<<"ball pos: "<<RectBoundingNextBallPos<<std::endl;
+  // std::cout<<"rect pos: "<<rect<<std::endl;
+  return rect.intersects(RectBoundingNextBallPos);
+}
+
+void BasicApp::drawLeftBat()
+{
+  gl::color(Color(1, 0, 0));  // red
```

Comment:
Better than a comment that one day might be outdated: self documenting code: `static const Color Red{1, 0, 0}`